sequencing-server/src/lib/codegen/CommandEDSLPreface.ts

```diff
@@ -522,39 +765,81 @@ export class Sequence implements SeqJson {
     }*/
 
     return (
-      `export default () =>\n` +
-      `${indent(`Sequence.new({`, 1)}\n` +
-      `${indent(`seqId: '${this.id}'`, 2)},\n` +
-      `${indent(`metadata: ${metadataString}`, 2)},\n` +
-      `${localsString.length !== 0 ? `${indent(`locals: ${localsString}`, 2)},\n` : ''}` +
-      `${parameterString.length !== 0 ? `${indent(`parameters: ${parameterString}`, 2)},\n` : ''}` +
-      `${commandsString.length !== 0 ? `${indent(`steps: ${commandsString}`, 2)},\n` : ''}` +
-      `${hardwareString.length !== 0 ? `${indent(`hardware_commands: ${hardwareString}`, 2)},\n` : ''}` +
-      `${immediateString.length !== 0 ? `${indent(`immediate_commands: ${immediateString}`, 2)},\n` : ''}` +
-      `${requestString.length !== 0 ? `${indent(`requests: ${requestString}`, 2)},\n` : ''}` +
-      `${indent(`});`, 1)}`
+        `export default () =>\n` +
+        `${indent(`Sequence.new({`, 1)}\n` +
+        `${indent(`seqId: '${this.id}'`, 2)},\n` +
+        `${indent(`metadata: ${metadataString}`, 2)},\n` +
+        `${localsString.length !== 0 ? `${indent(`locals: ${localsString}`, 2)},\n` : ''}` +
+        `${parameterString.length !== 0 ? `${indent(`parameters: ${parameterString}`, 2)},\n` : ''}` +
+        `${
+            commandsString.length !== 0 ? `${indent(`steps: ({ locals, parameters }) => (${commandsString}`, 2)}),\n` : ''
+        }` +
+        `${hardwareString.length !== 0 ? `${indent(`hardware_commands: ${hardwareString}`, 2)},\n` : ''}` +
+        `${immediateString.length !== 0 ? `${indent(`immediate_commands: ${immediateString}`, 2)},\n` : ''}` +
+        `${requestString.length !== 0 ? `${indent(`requests: ${requestString}`, 2)},\n` : ''}` +
+        `${indent(`});`, 1)}`
     );
   }
 
   // @ts-ignore : 'Args' found in JSON Spec
   public static fromSeqJson(json: SeqJson): Sequence {
+    const localNames = json.locals
+        ? [
+          json.locals[0].name,
```

Comment:
Why do we have to do this slice here? Something to do with preserving tuple order in the typechecker?

Reply:
The reason for the slice is that an empty array is not allowed for `locals/parameters` defined in the `SeqJSON` which is what `map` could generate. The same restriction applies to the `commands/steps` and `request` fields. If you are aware of an alternative method, I would be open to exploring it.

`locals : [VariableDeclaration, ...VariableDeclaration[]]`
`parameters : [VariableDeclaration, ...VariableDeclaration[]]`

Reply:
You should also be able to do
```
const localNames = json.locals !== undefined ? json.locals.map(local => local.name) : [];
```

Again, I might be missing something as I don't have everything in vscode right now.